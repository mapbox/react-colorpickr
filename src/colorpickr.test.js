import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ColorPickr from './colorpickr';
import '@testing-library/jest-dom/extend-expect';

describe('Colorpickr', () => {
  describe('renders', () => {
    let wrapper;
    const props = {
      onChange: jest.fn()
    };

    beforeEach(() => {
      wrapper = render(<ColorPickr {...props} />);
    });

    test('renders', () => {
      expect(wrapper.baseElement).toMatchSnapshot();
    });

    test('hex input returns value onChange', () => {
      const mockEvent = {
        target: {
          value: 'eeef'
        }
      };

      const input = wrapper.getByTestId('hex-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 93,
        r: 238,
        g: 238,
        b: 238,
        a: 1,
        hexInput: true,
        hex: 'eeef',
        mode: 'hsl',
        channel: 'h'
      });
    });

    test('invalid hex input does not fire onChange', () => {
      const mockEvent = {
        target: {
          value: 'eeeff'
        }
      };

      const input = wrapper.getByTestId('hex-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(0);
    });

    test('leading hash is permitted', () => {
      const mockEvent = {
        target: {
          value: '#333'
        }
      };

      const input = wrapper.getByTestId('hex-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 20,
        r: 51,
        g: 51,
        b: 51,
        a: 1,
        hexInput: true,
        hex: '333',
        mode: 'hsl',
        channel: 'h'
      });
    });

    test('hex input adjusts value onBlur', () => {
      const mockEvent = {
        target: {
          value: 'eeef'
        }
      };

      const input = wrapper.getByTestId('hex-input');
      fireEvent.blur(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 93,
        r: 238,
        g: 238,
        b: 238,
        a: 1,
        hexInput: true,
        hex: 'eeeeee',
        mode: 'hsl',
        channel: 'h'
      });
    });

    test('default mode is active', () => {
      const element = wrapper.getByTestId('mode-hsl');
      expect(element.checked).toEqual(true);
    });

    test('default reset action is present', () => {
      const element = wrapper.getByTestId('color-reset');
      expect(element).toBeTruthy();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });
  });

  describe('overrides', () => {
    let wrapper;
    let mockedInstance;

    const props = {
      onChange: jest.fn(),
      mounted: (instance) => (mockedInstance = instance)
    };

    beforeEach(() => {
      wrapper = render(<ColorPickr {...props} />);
    });

    test('mocked instance is called', () => {
      mockedInstance.overrideValue('red');
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 1,
        hexInput: false,
        hex: 'ff0000',
        mode: 'hsl',
        channel: 'h'
      });

      const element = wrapper.getByTestId('color-reset');
      expect(element.style.backgroundColor).toEqual('rgb(0, 0, 0)');
    });

    test('overrideValue with true as second argument sets overrides initialValue', () => {
      mockedInstance.overrideValue('red', true);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 1,
        hexInput: false,
        hex: 'ff0000',
        mode: 'hsl',
        channel: 'h'
      });

      const element = wrapper.getByTestId('color-reset');
      expect(element.style.backgroundColor).toEqual('rgb(255, 0, 0)');
    });

    afterEach(() => {
      jest.clearAllMocks();
    });
  });

  describe('readOnly', () => {
    const props = {
      onChange: jest.fn(),
      readOnly: true
    };

    test('renders', () => {
      const { baseElement, getByTestId, getByLabelText } = render(
        <ColorPickr {...props} />
      );
      expect(baseElement).toMatchSnapshot();
      expect(getByTestId('hex-input')).toHaveAttribute('readonly');
      expect(getByTestId('color-reset')).toHaveAttribute('disabled');
      expect(getByTestId('color-slider')).toHaveAttribute('disabled');
      expect(getByTestId('alpha-slider')).toHaveAttribute('disabled');
      expect(getByLabelText('α')).toHaveAttribute('readonly');
      expect(getByLabelText('h')).toHaveAttribute('readonly');
      expect(getByLabelText('s')).toHaveAttribute('readonly');
      expect(getByLabelText('l')).toHaveAttribute('readonly');
    });
  });

  describe('reset', () => {
    const props = {
      onChange: jest.fn(),
      reset: false
    };

    test('renders', () => {
      const { queryByTestId } = render(<ColorPickr {...props} />);
      expect(queryByTestId('color-reset')).toBeFalsy();
    });
  });

  describe('reset', () => {
    const props = {
      onChange: jest.fn(),
      alpha: false
    };

    test('renders', () => {
      const { queryByLabelText, queryByTestId } = render(
        <ColorPickr {...props} />
      );
      expect(queryByLabelText('α')).toBeFalsy();
      expect(queryByTestId('alpha-slider')).toBeFalsy();
    });
  });

  describe('hex value', () => {
    test('hex value remains long', () => {
      const props = {
        initialValue: '#33ffee',
        onChange: jest.fn()
      };

      const { getByTestId } = render(<ColorPickr {...props} />);
      const element = getByTestId('hex-input');
      expect(element.value).toEqual('33ffee');
    });

    test('hex value remains short', () => {
      const props = {
        initialValue: '#3fe',
        onChange: jest.fn()
      };

      const { getByTestId } = render(<ColorPickr {...props} />);
      const element = getByTestId('hex-input');
      expect(element.value).toEqual('3fe');
    });

    test('hex value maintains alpha onChange', () => {
      const props = {
        initialValue: 'hsla(100, 50%, 40%, 0.5)',
        onChange: jest.fn()
      };

      const { getByTestId } = render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: '#333'
        }
      };

      const input = getByTestId('hex-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 20,
        r: 51,
        g: 51,
        b: 51,
        a: 0.5,
        hexInput: true,
        hex: '333',
        mode: 'hsl',
        channel: 'h'
      });
    });
  });

  describe('modes', () => {
    test('rgb', () => {
      const props = {
        initialValue: 'rgba(0, 255, 255, 0.5)',
        onChange: jest.fn()
      };

      const { getByTestId } = render(<ColorPickr {...props} />);
      const input = getByTestId('mode-rgb');

      const mockEvent = {
        target: {
          value: 'rgb'
        }
      };

      userEvent.click(input);
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 180,
        s: 100,
        l: 50,
        r: 0,
        g: 255,
        b: 255,
        a: 0.5,
        hexInput: false,
        hex: '00ffff',
        mode: 'rgb',
        channel: 'h'
      });
    });

    test('hsl', () => {
      const props = {
        initialValue: 'hsla(180, 100%, 50%, 0.5)',
        mode: 'rgb',
        onChange: jest.fn()
      };

      const { getByTestId } = render(<ColorPickr {...props} />);
      const input = getByTestId('mode-hsl');

      const mockEvent = {
        target: {
          value: 'hsl'
        }
      };

      userEvent.click(input);
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 180,
        s: 100,
        l: 50,
        r: 0,
        g: 255,
        b: 255,
        a: 0.5,
        hexInput: false,
        hex: '00ffff',
        mode: 'hsl',
        channel: 'h'
      });
    });
  });
});
