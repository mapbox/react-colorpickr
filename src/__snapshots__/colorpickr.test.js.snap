// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colorpickr readOnly renders 1`] = `
<body>
  <div>
    <div
      class="colorpickr round inline-block bg-gray-faint w240 round px12 py12 txt-xs"
    >
      <div
        class="flex relative"
      >
        <div
          class="flex-child-no-shrink pb12 h120 w120 z1"
        >
          <div
            class="relative w-full h-full cursor-pointer events-none"
          >
            <div
              class="xy-control absolute z1 unselectable cursor-move"
              style="top: 100%; left: 0%;"
            />
            <noscript />
            <noscript />
            <noscript />
            <div
              class="absolute top right bottom left"
              style="background-color: rgb(255, 0, 0);"
            />
            <div
              class="absolute top right bottom left gradient-hue"
            />
            <noscript />
            <noscript />
          </div>
          <div
            class="slider colormode-slider colormode-slider-h"
          >
            <input
              data-testid="color-slider"
              disabled=""
              max="360"
              min="0"
              type="range"
              value="0"
            />
          </div>
          <div
            class="slider bg-tile bg-white"
          >
            <input
              data-testid="alpha-slider"
              disabled=""
              max="100"
              min="0"
              type="range"
              value="100"
            />
          </div>
        </div>
        <div
          class="flex-child-grow ml12"
        >
          <div
            class="toggle-group wmax-full mb12"
          >
            <label
              class="toggle-container w-full"
            >
              <input
                checked=""
                data-testid="mode-hsl"
                name="toggle"
                type="radio"
                value="hsl"
              />
              <div
                class="toggle toggle--s toggle--gray"
              >
                HSL
              </div>
            </label>
            <label
              class="toggle-container w-full"
            >
              <input
                data-testid="mode-rgb"
                name="toggle"
                type="radio"
                value="rgb"
              />
              <div
                class="toggle toggle--s toggle--gray"
              >
                RGB
              </div>
            </label>
          </div>
          <div>
            <div
              class="mb3 flex is-active"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  checked=""
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="h"
                >
                  h
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono"
                  id="h"
                  max="360"
                  min="0"
                  readonly=""
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="mb3 flex"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="s"
                >
                  s
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono"
                  id="s"
                  max="100"
                  min="0"
                  readonly=""
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="mb3 flex"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="l"
                >
                  l
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono"
                  id="l"
                  max="100"
                  min="0"
                  readonly=""
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
          <div
            class="mb3"
          >
            <div
              class="flex-child-grow relative"
            >
              <label
                class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                for="α"
              >
                α
              </label>
              <input
                class="w-full pl18 pr3 input input--s txt-mono"
                id="α"
                max="100"
                min="0"
                readonly=""
                step="1"
                type="number"
                value="100"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex mt6"
      >
        <div
          class="flex-child-no-shrink grid w120 round overflow-hidden"
        >
          <div
            class="bg-tile bg-white col col--auto h24 border-r border--gray-faint"
          >
            <button
              aria-disabled="true"
              aria-label="Reset color"
              class="w-full h-full txt-bold align-center color-transparent color-white-on-hover transition"
              data-testid="color-reset"
              disabled=""
              style="background-color: rgb(0, 0, 0);"
              title="Reset color"
              type="button"
            />
          </div>
          <div
            class="bg-tile bg-white col col--auto h24"
          >
            <div
              class="w-full h-full"
              style="background-color: rgb(0, 0, 0);"
            />
          </div>
        </div>
        <div
          class="flex-child-grow ml12 relative"
        >
          <label
            class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
          >
            #
          </label>
          <input
            class="w-full pl18 pr3 input input--s txt-mono"
            data-testid="hex-input"
            readonly=""
            type="text"
            value="000"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Colorpickr renders renders 1`] = `
<body>
  <div>
    <div
      class="colorpickr round inline-block bg-gray-faint w240 round px12 py12 txt-xs"
    >
      <div
        class="flex relative"
      >
        <div
          class="flex-child-no-shrink pb12 h120 w120 z1"
        >
          <div
            class="relative w-full h-full cursor-pointer"
          >
            <div
              class="xy-control absolute z1 unselectable cursor-move"
              style="top: 100%; left: 0%;"
            />
            <noscript />
            <noscript />
            <noscript />
            <div
              class="absolute top right bottom left"
              style="background-color: rgb(255, 0, 0);"
            />
            <div
              class="absolute top right bottom left gradient-hue"
            />
            <noscript />
            <noscript />
          </div>
          <div
            class="slider colormode-slider colormode-slider-h"
          >
            <input
              data-testid="color-slider"
              max="360"
              min="0"
              type="range"
              value="0"
            />
          </div>
          <div
            class="slider bg-tile bg-white"
          >
            <input
              data-testid="alpha-slider"
              max="100"
              min="0"
              type="range"
              value="100"
            />
          </div>
        </div>
        <div
          class="flex-child-grow ml12"
        >
          <div
            class="toggle-group wmax-full mb12"
          >
            <label
              class="toggle-container w-full"
            >
              <input
                checked=""
                data-testid="mode-hsl"
                name="toggle"
                type="radio"
                value="hsl"
              />
              <div
                class="toggle toggle--s toggle--gray"
              >
                HSL
              </div>
            </label>
            <label
              class="toggle-container w-full"
            >
              <input
                data-testid="mode-rgb"
                name="toggle"
                type="radio"
                value="rgb"
              />
              <div
                class="toggle toggle--s toggle--gray"
              >
                RGB
              </div>
            </label>
          </div>
          <div>
            <div
              class="mb3 flex is-active"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  checked=""
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="h"
                >
                  h
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono bg-white"
                  id="h"
                  max="360"
                  min="0"
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="mb3 flex"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="s"
                >
                  s
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono bg-white"
                  id="s"
                  max="100"
                  min="0"
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="mb3 flex"
            >
              <div
                class="flex-child-no-shrink flex flex--center-cross w24"
              >
                <input
                  class="cursor-pointer"
                  name=""
                  type="radio"
                />
              </div>
              <div
                class="flex-child-grow relative"
              >
                <label
                  class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                  for="l"
                >
                  l
                </label>
                <input
                  class="w-full pl18 pr3 input input--s txt-mono bg-white"
                  id="l"
                  max="100"
                  min="0"
                  step="1"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
          <div
            class="mb3"
          >
            <div
              class="flex-child-grow relative"
            >
              <label
                class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
                for="α"
              >
                α
              </label>
              <input
                class="w-full pl18 pr3 input input--s txt-mono bg-white"
                id="α"
                max="100"
                min="0"
                step="1"
                type="number"
                value="100"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex mt6"
      >
        <div
          class="flex-child-no-shrink grid w120 round overflow-hidden"
        >
          <div
            class="bg-tile bg-white col col--auto h24 border-r border--gray-faint"
          >
            <button
              aria-label="Reset color"
              class="w-full h-full txt-bold align-center color-transparent color-white-on-hover transition"
              data-testid="color-reset"
              style="background-color: rgb(0, 0, 0);"
              title="Reset color"
              type="button"
            >
              Reset
            </button>
          </div>
          <div
            class="bg-tile bg-white col col--auto h24"
          >
            <div
              class="w-full h-full"
              style="background-color: rgb(0, 0, 0);"
            />
          </div>
        </div>
        <div
          class="flex-child-grow ml12 relative"
        >
          <label
            class="absolute top left bottom pl6 flex flex--center-cross color-gray-light txt-bold"
          >
            #
          </label>
          <input
            class="w-full pl18 pr3 input input--s txt-mono bg-white"
            data-testid="hex-input"
            type="text"
            value="000"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
